aten/src/ATen/native/cuda/BatchLinearAlgebraLib.cu

```diff
@@ -0,0 +1,132 @@
+  const int batch_size = cuda_int_cast(batchCount(self), "batchCount");
+  const int n = cuda_int_cast(self.size(-2), "self.size(-2)");
+
+  auto self_data = self.data_ptr<scalar_t>();
+  auto self_mat_stride = matrixStride(self);
+  auto self_inv_data = self_inv.data_ptr<scalar_t>();
+  auto self_inv_mat_stride = matrixStride(self_inv);
+
+  Tensor self_array = at::empty({batch_size}, at::device(at::kCPU).dtype(at::kLong));
+  Tensor self_inv_array = at::empty({batch_size}, at::device(at::kCPU).dtype(at::kLong));
+  scalar_t** p_self_array = reinterpret_cast<scalar_t**>(self_array.data_ptr());
+  scalar_t** p_self_inv_array = reinterpret_cast<scalar_t**>(self_inv_array.data_ptr());
+
+  // Set up the created arrays
+  for (int64_t i = 0; i < batch_size; i++) {
+    p_self_array[i] = &self_data[i * self_mat_stride];
+    p_self_inv_array[i] = &self_inv_data[i * self_inv_mat_stride];
+  }
+
+  auto& allocator = *::c10::cuda::CUDACachingAllocator::get();
+
+  if (use_loop_launch(batch_size, n)) {
+    int* p_infos = infos.data_ptr<int>();
+    auto main_stream = at::cuda::getCurrentCUDAStream();
+
+    for (int64_t i = 0; i < batch_size; i++) {
+      auto stream = at::cuda::getStreamFromPool();
+      at::cuda::CUDAStreamGuard guard(stream);
+
+      at::cuda::CUDAEvent can_start;
+      can_start.record(main_stream);
+      can_start.block(main_stream);
+
+      int* pivot = reinterpret_cast<int*>(allocator.allocate(sizeof(int) * n).get());
+      _apply_single_inverse_helper<scalar_t>(
+        p_self_array[i], p_self_inv_array[i], pivot, p_infos + i, n);
+
+      at::cuda::CUDAEvent finished;
+      finished.record(stream);
+      finished.block(main_stream);
+    }
+  } else {
+    self_array = self_array.cuda();
+    self_inv_array = self_inv_array.cuda();
+    int* ipiv_array = reinterpret_cast<int*>(allocator.allocate(sizeof(int)*batch_size*n).get());
+
+    at::cuda::blas::getrfBatched<scalar_t>(n, n, reinterpret_cast<scalar_t**>(self_array.data_ptr()), n,
+      ipiv_array, infos.data_ptr<int>(), batch_size);
+
+    at::cuda::blas::getriBatched<scalar_t>(n, n, reinterpret_cast<scalar_t**>(self_array.data_ptr()), n,
+      ipiv_array, infos.data_ptr<int>(), batch_size, reinterpret_cast<scalar_t**>(self_inv_array.data_ptr()));
+  }
+}
+
+template <typename scalar_t>
+static void apply_single_inverse_lib(const Tensor& self, Tensor& self_inv, int64_t& info) {
+  int n = cuda_int_cast(self.size(-2), "self.size(-2)");
+
+  Tensor ipiv = at::empty({n}, self.options().dtype(at::kInt));
+  Tensor info_tmp = at::zeros({1}, self.options().dtype(at::kInt));
+
+  _apply_single_inverse_helper<scalar_t>(
+    self.data_ptr<scalar_t>(), self_inv.data_ptr<scalar_t>(), ipiv.data_ptr<int>(), info_tmp.data_ptr<int>(), n);
+
+  info = info_tmp.item<int>();
+}
+
+Tensor _inverse_helper_cuda_lib(const Tensor& self) {
+  Tensor self_working_copy = cloneBatchedColumnMajor(self);
```

Comment:
this would unconditionally make a copy of the source, but is it always necessary? If matrix is already column major, and batch dimensions are contiguous, then copy is not needed?

Reply:
I think it would be necessary. After LU decomposition cusolver-getrf and cublas-batched-getrf, the passed-in matrix are overwritten with LU factorization results.